colossalai/moe/_operation.py

```diff
@@ -138,8 +138,66 @@ def forward(
     @staticmethod
     def backward(ctx: Any, *grad_outputs) -> Tuple[Tensor, None, None]:
         return (
-            AllToAll.forward(None, grad_outputs[0], ctx.comm_grp)[0],
+            AllToAll.forward(None, grad_outputs[0], ctx.comm_grp, False)[0],
+            None,
             None,
+        )
+
+
+class HierarchicalAllToAll(torch.autograd.Function):
+
+    @staticmethod
+    def forward(
+        ctx: Any,
+        inputs: Tensor,
+        groups: Tuple[ProcessGroup],
+    ) -> Tensor:
+        """
+        Returns:
+            outputs: Tensor
+        """
+        if ctx is not None:
+            ctx.comm_grps = groups
+        intra_node_group, inter_node_group = groups
+
+        local_world_size = dist.get_world_size(intra_node_group)
+        num_group = dist.get_world_size(inter_node_group) if inter_node_group is not None else 1
+        world_size = local_world_size * num_group
+        src_rank = dist.get_process_group_ranks(intra_node_group)[0]
+        outputs = torch.empty_like(inputs)
+
+        if dist.get_rank() == src_rank:
+            # intra-node gather
+            intra_output = [torch.empty_like(inputs) for _ in range(local_world_size)]
+            dist.gather(inputs, intra_output, dst=src_rank, group=intra_node_group)
+
+            intra_output = [v.chunk(world_size, dim=0) for v in intra_output]
+            intra_output = torch.cat(sum(zip(*intra_output), ()))
+
+            # inter-node all-to-all
```

Comment:
After inter node allgather, can we squeeze its empty space before alltoall? Then restore its layout when alltoall send back.

Reply:
You can also can add a todo here and implement in next pr

Reply:
38b7bbd marks it as TODO for now.